lib/commanded/aggregates/aggregate.ex

```diff
@@ -77,32 +77,57 @@ defmodule Commanded.Aggregates.Aggregate do
+  defp execute_command(command, handler, %Aggregate{aggregate_state: %{version: version} = aggregate_state} = state, retries) when retries > 0 do
     expected_version = version
 
-    aggregate_state = handler.handle(aggregate_state, command)
+    with {:ok, aggregate_state} <- handle_command(handler, aggregate_state, command),
+         {:ok, aggregate_state} <- persist_events(aggregate_state, expected_version)
+      do {:ok, %Aggregate{state | aggregate_state: aggregate_state}}
+    else
+      {:error, :wrong_expected_version} -> retry_command(command, handler, state, retries - 1)
+      {:error, _reason} = reply -> {reply, state}
+    end
+  end
+
+  defp execute_command(command, _handler, _aggregate, retries) when retries == 0 do
+    Logger.warn(fn -> "failed to execute command #{inspect command} after #{@command_retries}, returning an error" end)
+
+    {:error, :command_failed_after_retrying}
+  end
 
-    case persist_events(aggregate_state, expected_version) do
-      :ok -> %Aggregate{state | aggregate_state: %{aggregate_state | pending_events: []}}
-      {:error, :wrong_expected_version} ->
-        Logger.error(fn -> "failed to persist events for aggregate #{aggregate_uuid} due to wrong expected version" end)
+  # retry a command when an expected version mismatch error is encountered after reloading the aggregate's events
+  defp retry_command(command, handler, %Aggregate{} = state, retries) do
+    # reload aggregate's events
+    state = load_events(state)
 
-        # reload aggregate's events
-        state = load_events(state)
+    # retry command
+    execute_command(command, handler, state, retries)
+  end
 
-        # retry command
-        execute_command(command, handler, state, retries - 1)
+  defp handle_command(handler, aggregate_state, command) do
+    case handler.handle(aggregate_state, command) do
+      {:error, reason} = reply ->
+        Logger.debug(fn -> "failed to handle command due to: #{inspect reason}" end)
+        reply
+      aggregate_state -> {:ok, aggregate_state}
     end
   end
 
   # no pending events to persist, do nothing
-  defp persist_events(%{pending_events: []}, _expected_version), do: :ok
+  defp persist_events(%{pending_events: []} = aggregate_state, _expected_version), do: {:ok, aggregate_state}
 
-  defp persist_events(%{uuid: aggregate_uuid, pending_events: pending_events}, expected_version) do
+  defp persist_events(%{uuid: aggregate_uuid, pending_events: pending_events} = aggregate_state, expected_version) do
     correlation_id = UUID.uuid4
     event_data = Mapper.map_to_event_data(pending_events, correlation_id)
 
-    EventStore.append_to_stream(aggregate_uuid, expected_version, event_data)
+    case EventStore.append_to_stream(aggregate_uuid, expected_version, event_data) do
+      :ok ->
+        # clear pending events after successful store
+        {:ok, %{aggregate_state | pending_events: []}}
+      {:error, :wrong_expected_version} = reply ->
```

Comment:
Are you suggesting to remove the `{:error, :wrong_expected_version}` match and only match on :ok`.  This would crash the aggregate process if persisting the pending events fails due to version mismatch. That would follow the 'let it crash' idiom.

I haven't yet thought too much about retrying command handling on failure. I'd like this to be client configurable (e.g. 5 retries, with exponential back-off).